mpcs/src/basefold/structure.rs

```diff
@@ -61,52 +64,73 @@ pub struct BasefoldCommitmentWithWitness<E: ExtensionField>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
-    pub(crate) pi_d_digest: Digest<E>,
+    pub(crate) commit: Digest<E>,
     pub(crate) codeword: MerkleTree<E::BaseField>,
-    pub(crate) polynomials_bh_evals: Vec<ArcMultilinearExtension<'static, E>>,
-    pub(crate) num_vars: usize,
-    pub(crate) is_base: bool,
-    pub(crate) num_polys: usize,
+
+    pub(crate) log2_max_codeword_size: usize,
+    pub(crate) trivial_proofdata: BTreeMap<usize, (Digest<E>, MerkleTree<E::BaseField>)>,
+    // poly groups w.r.t circuit index
+    pub(crate) polys: BTreeMap<usize, Vec<ArcMultilinearExtension<'static, E>>>,
+    // keep codeword index w.r.t circuit index
+    pub circuit_codeword_index: BTreeMap<usize, usize>,
 }
 
 impl<E: ExtensionField> BasefoldCommitmentWithWitness<E>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
+    pub fn new(
+        commit: Digest<E>,
+        codeword: MerkleTree<E::BaseField>,
+        polys: BTreeMap<usize, Vec<ArcMultilinearExtension<'static, E>>>,
+        trivial_proofdata: BTreeMap<usize, (Digest<E>, MerkleTree<E::BaseField>)>,
+        circuit_codeword_index: BTreeMap<usize, usize>,
+    ) -> Self {
+        let mmcs = poseidon2_merkle_tree::<E>();
+        // size = height * 2 because we concat pi[left]/pi[right] under same row index
```

Comment:
nitpick: we need to clarify the meaning of `pi` here?

Reply:
Right and thanks for the check.

I removed the comment here besides switch to use `rmm.values.len()` so we dont need to explain why heightx2 here :)